rfc/0003-indexmanagement.md

```diff
@@ -0,0 +1,156 @@
+# Meta
+
+ - RFC Name: Index Management
+ - RFC ID: 0003-indexmanagement
+ - Start Date: 2015-12-14
+ - Owner: Simon Baslé (@simonbasle)
+ - Current Status: Draft
+
+# Summary
+The aim of the "`Index Management`" RFC is to offer a simplified API in the `BucketManager` to list, create and drop indexes, for the majority of use cases. In this sense, it is focused on GSI indexes and assumes a few other givens. For cases that don't fit in this picture, the fallback for the user is to craft appropriate `N1QL` queries (eg. in Java using the `Index` fluent API).
+
+# Motivation
+The goal is to abstract away regularly needed queries relative to indexes and provide a simple API. This way the user can focus on his business related `N1QL` queries, without having to think twice about the boilerplate of creating indexes.
+
+# General Design
+The API will be added to the `BucketManager` class of each SDK. In its blocking form, the API is made up of at least **6** base methods:
+ - `listIndex`: lists all the indexes relating to the current `Bucket` using the `system:indexes` N1QL view.
+ - `createPrimaryIndex`: create a primary index on the current Bucket.
+ - `createIndex`: create a secondary GSI index on the current Bucket.
+ - `dropPrimaryIndex`: drop the current Bucket's primary index.
+ - `dropIndex`: drop a specific secondary index of the current Bucket.
+ - `buildDeferredIndexes`: trigger the build of indexes that were created in a deferred fashion (see below).
+
+Additionally it is open for discussion wether or not this RFC should also include the following ***2*** methods in a cross-SDK fashion, or if they should be optional:
+ - `watchIndex`: poll the `system:indexes` N1QL view until a given index changes its status to "online" (or a maximum amount of time has passed).
+ - `watchIndexes`: poll the `system:indexes` N1QL view until a given list of index all become "online" (or a maximum amount of time has passed).
+
+
+The underlying design would be to rely on existing `N1QL` verbs and, for the listing and polling part, to use the `system:indexes` keyspace. Each method will internally create and execute a specific N1QL query according to its parameter and the name of the current Bucket.
+
+Each method will offer a bit of tuning with a couple of parameters, so let's detail a more complete signature for each one.
+
+## Listing Indexes
+Signature:
+```java
+List<IndexInfo> listIndexes()
```

Comment:
I used the `java` fenced code block so that the result is syntax highlighted and more readable, but I tried to use things that are generic and self-explanatory enough...